### ISSUE TITLE
KDE, unable to read dark/light color scheme preference

### ISSUE THREAD

**Issue by peterhoeg** [NixOS/nixpkgs#202810]
### Describe the bug

Previously, it used to be possible to query the user's dark/light color scheme preference via dbus as follows:

```sh
qdbus org.freedesktop.portal.Desktop /org/freedesktop/portal/desktop org.freedesktop.portal.Settings.Read "org.freedesktop.appearance" "color-scheme"
```

I have been using something similar to this from emacs (basically subscribing for changes to the relevant property but that difference is immatterial here) in order to have it change colour automatically.

But suddenly (I don't know when), querying dbus always returns 0 which indicates "no preference" rather than 1 or 2.

### Steps To Reproduce
1. set the color scheme in KDE to dark
2. run the command above and notice it returns 0
3. set the color scheme in KDE to light
4. run the command above and notice it returns 0

### Expected behavior
The returned value should indicate the dark/light preference.

### Notify maintainers
Cc: @NixOS/qt-kde

### Metadata
Latest unstable but probably stopped working within the last couple of weeks. Problem persists on the generations available locally.

**Comment by peterhoeg** [comment]
And today it's working as normal again.

I have *no* idea what caused this. I saw it on multiple machines with a couple of reboots but as of today, it's all good. :shrug:

**Comment by peterhoeg** [comment]
And this problem is back. These days, it isn't working most of the time, but then suddenly it does.

Super weird.

**Comment by samueldr** [comment]
I don't know if it will be of any help, but dump `systemctl --user` and/or `systemctl --user status` from when it works, and when it doesn't. Might be a pain to `diff` (mechanically), but maybe there will be clues about a service that failed to launch or exited abnormally?

**Comment by peterhoeg** [comment]
`xdg-desktop-portal' is the culprit.

I’m guessing it’s a race condition of some kind because restarting it with `systemctl --user restart xdg-desktop-portal' makes it work as expected.

**Comment by peterhoeg** [comment]
So just to elaborate - there were no crashed/failed services, but I used `busctl` to see what service was responsible for `org.freedesktop.portal.Desktop` which turned out to be `xdg-desktop-portal` (I probably should have been able to guess that).

I then rebooted the machine a few times and every time, the current scheme couldn't be read after start-up but restarting that particular service was enough to make things work.

I'm guessing that whatever component is responsible for providing the portal service with the settings only comes up *after* the portal service is running and that messes it up.

**Comment by peterhoeg** [comment]
I had `xdg-desktop-portal-gtk` added as well to the environment which caused the `qdbus` invocation mentioned above to return 0 in some cases. I'm guessing because the gtk backend was queried before the kde backend.

Now that I have removed it, I no longer get a 0 in response, but instead it doesn't change when the I change from light to dark and vice versa. Instead both `xdg-desktop-portal` and `plasma-xdg-desktop-portal-kde` need to be restarted in order to be able to "see" the scheme.

Basically this is what happens:

1. session starts with light theme
2. `qdbus` returns light (value 2)
3. change the scheme from light to dark via system settings
4. `qdbus` returns light (value 2)
5. restart the portal services
6. `qdbu` returns dark (value 1)

**Comment by keenanweaver** [comment]
I'm on the unstable channel and am experiencing this. Adding this as a login script in the Plasma options works sufficiently as a workaround, at least for me:

``` bash
#!/usr/bin/env bash
systemctl --user restart xdg-desktop-portal plasma-xdg-desktop-portal-kde
```

**Comment by peterhoeg** [comment]
But don't you need to restart those services again when switching from dark to light or vice versa?

**Comment by alnmy** [comment]
Experiencing same issue, https://github.com/NixOS/nixpkgs/issues/202810#issuecomment-1532008182 works as a solution for me.
